### PR TITLE
fixed the a-and-b-same-order algorithm

### LINKED ISSUE #34: Wrong algorihtm for a-and-b-same-order in comparison function
I realise this is the same as we had before, but I'm actually not sure this is correct, I think the same-order check should only look at common elements. Example:
```
A = { 
  names: ["chr1", "chr2", "chr3", "M"],
  lengths: [1, 2, 2, 3],
  sequences: [...]
}
B = {
  names: ["chr1", "chr2", "chr3"],
  lengths: [1, 2, 2],
  sequences: [...]
}
```
Then `A.lengths != B.lengths` but their common elements are in the same order. Similarly for names (and presumably sequences if I bothered to write digests there).

You can make a separate issue for this and do it later, it should be straightforward by re-using the logic of `getCommonElementsCount`.

_Originally posted by @apriltuesday in https://github.com/EBIvariation/eva-seqcol/pull/33#discussion_r1297420234_

### PR DESCRIPTION
**Pull request by waterflow80**
closes #34

### CODE REVIEW

File: src/main/java/uk/ac/ebi/eva/evaseqcol/service/SeqColService.java [hunk 1]
```diff
@@ -258,32 +259,55 @@ public SeqColComparisonResultEntity compareSeqCols(
...
+    /**
+     * Check whether the array of elements of elementsA are in the same order as the ones in elementsB.
+     * Example 1:
+     *      A = ["ch1", "B", "ch2", "ch3"],
+     *      B = ["ch1", "A", "ch2", "ch3"],
+     *      Common = ["ch1", "ch2", "ch3"] # Common elements between A and B
+     *      ==> A.indexOf("ch1") == B.indexOf("ch1"), A.indexOf("ch2") == B.indexOf("ch2"), A.indexOf("ch3") == B.indexOf("ch3")
+     *      ==> Same order elements
+     *      */
+    public boolean check_A_And_B_Same_Order(List<String> elementsA, List<String> elementsB) {
```
Comment (sundarvenkata-EBI) [inline src/main/java/uk/ac/ebi/eva/evaseqcol/service/SeqColService.java:297]: Suggest including [this link](https://github.com/ga4gh/seqcol-spec/blob/master/docs/decision_record.md#same-order-specification) in the comment

Comment (waterflow80) [reply src/main/java/uk/ac/ebi/eva/evaseqcol/service/SeqColService.java:297]: Absolutely

File: src/main/java/uk/ac/ebi/eva/evaseqcol/entities/SeqColComparisonResultEntity.java [hunk 2]
```diff
@@ -52,4 +52,10 @@ public void putIntoElements(String elementName,String key, Object value) {
         SortedMap<String, Object> elementsMap = elements.get(elementName);
         elementsMap.put(key, value);
     }
+
+    /**
+     * Return the value of the given attributeName under the given elementName*/
+    public Boolean getElement(String elementName, String attributeName) {
+        return (Boolean) elements.get(elementName).get(attributeName);
```
Comment (apriltuesday) [inline src/main/java/uk/ac/ebi/eva/evaseqcol/entities/SeqColComparisonResultEntity.java:59]: The values can be Integer or Boolean, so should probably check before casting... Also, I don't see where we actually use this method.

File: src/main/java/uk/ac/ebi/eva/evaseqcol/service/SeqColService.java [hunk 3]
```diff
@@ -258,32 +259,64 @@ public SeqColComparisonResultEntity compareSeqCols(
...
+     *      ==> Same order elements
+     * NOTE: Assuming that the method List.retainAll() preserves the order in the original list (no counterexample at the moment)
+     * @see "https://github.com/ga4gh/seqcol-spec/blob/master/docs/decision_record.md#same-order-specification" */
+    public boolean check_A_And_B_Same_Order(List<String> elementsA, List<String> elementsB) {
+        LinkedList<String> elementsALocal = new LinkedList<>(elementsA);
+        LinkedList<String> elementsBLocal = new LinkedList<>(elementsB);
+        List<String> commonElements = getCommonElementsDistinct(elementsALocal, elementsBLocal);
+        elementsALocal.retainAll(commonElements); // Leaving only the common elements (keeping the original order to check)
+        elementsBLocal.retainAll(commonElements); // Leaving only the common elements (keeping the original order to check)
+        for (String element: commonElements) {
```
Comment (apriltuesday) [inline src/main/java/uk/ac/ebi/eva/evaseqcol/service/SeqColService.java:310]: I think once you've done `retainAll` (assuming the order is preserved), you should just be able to rely on list equality for the rest - i.e. just `return elementsALocal.equals(elementsBLocal);`

File: src/test/java/uk/ac/ebi/eva/evaseqcol/service/SeqColServiceTest.java [hunk 4]
```diff
@@ -90,4 +92,24 @@ void unbalancedDuplicatesPresentTest() {
...
+        List<String> listB1 = new ArrayList<>(Arrays.asList("chr1", "chr2", "chr3"));
+        List<String> listB2 = new ArrayList<>(Arrays.asList("1", "2", "2"));
+        List<String> listB3 = new ArrayList<>(Arrays.asList("ch1", "A", "ch2", "ch3"));
+        List<String> listB4 = new ArrayList<>(Arrays.asList("A", "ch1", "ch2", "ch3"));
+
+        assertTrue(seqColService.check_A_And_B_Same_Order(listA1, listB1));
+        assertTrue(seqColService.check_A_And_B_Same_Order(listA2, listB2));
+        assertTrue(seqColService.check_A_And_B_Same_Order(listA3, listB3));
+        assertTrue(seqColService.check_A_And_B_Same_Order(listA4, listB4));
+
```
Comment (apriltuesday) [inline src/test/java/uk/ac/ebi/eva/evaseqcol/service/SeqColServiceTest.java:112]: Might want to add a negative example, e.g. `A = [1,2,2]` and `B = [2,1,2]` should be false.

File: src/main/java/uk/ac/ebi/eva/evaseqcol/service/SeqColService.java (hunk 3 above)
Comment (waterflow80) [reply src/main/java/uk/ac/ebi/eva/evaseqcol/service/SeqColService.java:310]: True, I'll do that.

**Review by apriltuesday — APPROVED** [review]

**Review by sundarvenkata-EBI — APPROVED** [review]

**Review by tcezard — APPROVED** [review]

**Pull request merged**